src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -703,58 +704,32 @@ private BoundStatement BindDeclarationStatementParts(LocalDeclarationStatementSy
-                                                                  MessageID.IDS_Disposable);
-
-            // we failed if we didn't find a void returning method; try searching extension methods
-            // PROTOTYPE: what happens if there is a no arg int returning method, but a params object[] void returning. Currently, overload resolution would choose the int return, and we'd fail here.
-            if (disposeMethod is null || !disposeMethod.ReturnsVoid)
-            {
-                lookupResult.Clear();
-
-                var extDisposeMethod = FindPatternExtensionMethod(typeExpr,
-                                                                  WellKnownMemberNames.DisposeMethodName,
-                                                                  syntaxNode,
-                                                                  diagnostics,
-                                                                  MessageID.IDS_Disposable);
-
-                // if we still can't satisfy the lookup via extension method, report the diagnostics from both lookups
-                if (extDisposeMethod is null || !extDisposeMethod.ReturnsVoid)
-                {
-                    diagnostics.AddRange(initialDiagnostics);
-                    if (disposeMethod?.ReturnsVoid == false)
-                    {
-                        ReportPatternWarning(diagnostics, typeExpr.Type, disposeMethod, syntaxNode, MessageID.IDS_Disposable);
-                    }
-                    if (extDisposeMethod?.ReturnsVoid == false)
-                    {
-                        ReportPatternWarning(diagnostics, typeExpr.Type, extDisposeMethod, syntaxNode, MessageID.IDS_Disposable);
-                    }
-                    extDisposeMethod = null;
-                }
+        /// <param name="expr">Expression on which to perform lookup</param>
+        /// <param name="syntaxNode">The syntax node to perform lookup on</param>
+        /// <param name="diagnostics">Populated with invocation errors, and warnings of near misses</param>
+        /// <returns>The <see cref="MethodSymbol"/> of the Dispose method if one is found, otherwise null.</returns>
+        internal MethodSymbol TryFindDisposePatternMethod(BoundExpression expr, SyntaxNode syntaxNode, DiagnosticBag diagnostics)
+        {
+            if(expr?.Type is null)
+            {
+                return null;
+            }
 
-                disposeMethod = extDisposeMethod;
+            var result = FindPatternMethodRelaxed(expr,
+                                                  WellKnownMemberNames.DisposeMethodName,
+                                                  syntaxNode,
+                                                  diagnostics,
+                                                  out var disposeMethod);
+
+            
+            if (disposeMethod?.ReturnsVoid == false || result == PatternLookupResult.NotAMethod)
```

Comment:
>result == PatternLookupResult.NotAMethod [](start = 55, length = 40)

What about other possible result values? (such as `PatternLookupResult.NotCallable`)
Is it better to report such cases as pattern mismatch or no conversion to IDisposable?

Reply:
In all cases other than NotAMethod, the bindmethodgroup will add applicable errors. This case handles when Dispose wasn't a method (so we didn't try and bind), or when x.Dispose() was a valid call, but didn't fit the pattern: I'm clarifying the logic in the speclet.

Reply:
For such situations in other contexts, I used enums with 3 values: Success, Failed, FailedNotReported. Consider whether that is applicable here, or if we care about the more detailed results with specific semantics.

---
In reply to: [225302376](https://github.com/dotnet/roslyn/pull/30437#discussion_r225302376) [](ancestors = 225302376)